torch/utils/tensorboard/writer.py

```diff
@@ -0,0 +1,731 @@
+              When logging crashes at step :math:`T+X` and restarts at step :math:`T`, any events
+              whose global_step larger or equal to :math:`T` will be purged and hidden from TensorBoard.
+              Note that the resumed experiment and crashed experiment should have the same ``log_dir``.
+            filename_suffix (string):
+              Every event file's name is suffixed with suffix. Example: ``SummaryWriter(filename_suffix='.123')``
+              More details on event filename construction in
+              tensorboard.summary.writer.event_file_writer.EventFileWriter.
+            kwargs: extra keyword arguments for FileWriter (e.g. 'flush_secs'
+              controls how often to flush pending events). For more arguments
+              please refer to docs for 'tf.summary.FileWriter'.
+        """
+        if not log_dir:
+            import socket
+            from datetime import datetime
+            current_time = datetime.now().strftime('%b%d_%H-%M-%S')
+            log_dir = os.path.join(
+                'runs', current_time + '_' + socket.gethostname() + comment)
+        self.log_dir = log_dir
+        self.kwargs = kwargs
+
+        # Initialize the file writers, but they can be cleared out on close
+        # and recreated later as needed.
+        self.file_writer = self.all_writers = None
+        self.get_file_writer()
+
+        # Create default bins for histograms, see generate_testdata.py in tensorflow/tensorboard
+        v = 1E-12
+        buckets = []
+        neg_buckets = []
+        while v < 1E20:
+            buckets.append(v)
+            neg_buckets.append(-v)
+            v *= 1.1
+        self.default_bins = neg_buckets[::-1] + [0] + buckets
+        self.scalar_dict = {}
+
+    def __append_to_scalar_dict(self, tag, scalar_value, global_step,
+                                timestamp):
+        """This adds an entry to the self.scalar_dict datastructure with format
+        {writer_id : [[timestamp, step, value], ...], ...}.
+        """
+        if tag not in self.scalar_dict.keys():
+            self.scalar_dict[tag] = []
+        self.scalar_dict[tag].append(
+            [timestamp, global_step, float(make_np(scalar_value))])
+
+    def _check_caffe2_blob(self, item):
+        """
+        Caffe2 users have the option of passing a string representing the name of
+        a blob in the workspace instead of passing the actual Tensor/array containing
+        the numeric values. Thus, we need to check if we received a string as input
+        instead of an actual Tensor/array, and if so, we need to fetch the Blob
+        from the workspace corresponding to that name. Fetching can be done with the
+        following:
+
+        from caffe2.python import workspace (if not already imported)
+        workspace.FetchBlob(blob_name)
+        workspace.FetchBlobs([blob_name1, blob_name2, ...])
+        """
+        return isinstance(item, six.string_types)
+
+    def get_file_writer(self):
```

Comment:
Under what circumstances would a user want to get the underlying file writer? Can this be made a strictly internal implementation detail?

Reply:
Changing.